### PR TITLE
[Rule Tuning] Add Sequence for OAuth Authorization to Custom App - Google Workspace

### PR DESCRIPTION
**Pull request by terrancedejesus**
## Issues
* https://github.com/elastic/security-team/issues/5915

## Summary
This rule was originally created in https://github.com/elastic/detection-rules/pull/2627. Originally it leveraged EQL for this tuning opportunity to sequence with new Google Workspace token (OAuth) logs. This rule not identifies when specific Google Workspace resources are copied from an external drive to the drive of a user in the organization. The Google Workspace resource is limited to specific types that can contain container-bound scripts.

With this sequence addition, we identify when a backend application (Apps Script) is granted access/permission via OAuth by the same user. 

Theoretically, this will identify when a Google Workspace user receives an email with a phishing link that copies a malicious document to their drive and contains a container-bound script. The container-bound script would execute and thus ask for permission to specific scopes in Google Workspace via OAuth. In this scenario, the user would grant permission.

This is very uncommon but we have additional room to tune with `google_workspace.token.scope.value` where we can focus on `write` or `deploy` permissions if need be.

<details>
<summary>Screenshots</summary>

Screenshot 2023-03-29 at 9 51 47 PM

</details>

<details>
<summary>Example Data</summary>

Copied Document.

```sql
{
  "_index": ".ds-logs-google_workspace.drive-default-2023.03.22-000009",
  "_id": "/vuQ/a+yErSEoiSvycXRGlI5JUw=",
  "_score": 1,
  "fields": {
    "elastic_agent.version": [
      "8.6.2"
    ],
    "event.category": [
      "file"
    ],
    "google_workspace.drive.visibility": [
      "private"
    ],
    "source.user.email": [
      "terrance@dejesusarcheology.com"
    ],
    "cloud.availability_zone": [
      "us-east1-b"
    ],
    "source.user.name.text": [
      "terrance"
    ],
    "source.geo.region_name": [
      "Ohio"
    ],
    "google_workspace.drive.billable": [
      true
    ],
    "source.ip": [
      "2603:6010:ed00:b789:1860:feb:feee:41e8"
    ],
    "agent.name": [
      "ubuntu-server-tdejesus"
    ],
    "source.geo.region_iso_code": [
      "US-OH"
    ],
    "event.agent_id_status": [
      "verified"
    ],
    "google_workspace.event.type": [
      "access"
    ],
    "google_workspace.drive.is_encrypted": [
      false
    ],
    "event.kind": [
      "event"
    ],
    "source.geo.city_name": [
      "Martins Ferry"
    ],
    "user.id": [
      "115903088752625509360"
    ],
    "input.type": [
      "httpjson"
    ],
    "google_workspace.drive.file.owner.email": [
      "terrance@dejesusarcheology.com"
    ],
    "data_stream.type": [
      "logs"
    ],
    "related.user": [
      "terrance"
    ],
    "tags": [
      "forwarded",
      "google-workspace-drive"
    ],
    "event.provider": [
      "drive"
    ],
    "cloud.provider": [
      "gcp"
    ],
    "cloud.machine.type": [
      "e2-medium"
    ],
    "google_workspace.drive.file.id": [
      "1rdRwX4S-yvUOsRKa_AonklhEhn8yICSlVQH1dxts1ow"
    ],
    "cloud.service.name": [
      "GCE"
    ],
    "file.type": [
      "file"
    ],
    "agent.id": [
      "f4011165-a1b8-4d8e-b902-56ba8835cc28"
    ],
    "ecs.version": [
      "8.6.0"
    ],
    "event.created": [
      "2023-03-29T20:15:36.445Z"
    ],
    "organization.id": [
      "C00qtspd5"
    ],
    "agent.version": [
      "8.6.2"
    ],
    "file.owner": [
      "terrance"
    ],
    "google_workspace.drive.file.owner.is_shared_drive": [
      false
    ],
    "google_workspace.drive.actor_is_collaborator_account": [
      false
    ],
    "source.user.name": [
      "terrance"
    ],
    "source.as.number": [
      10796
    ],
    "google_workspace.drive.copy_type": [
      "external"
    ],
    "user.name": [
      "terrance"
    ],
    "source.geo.location": [
      {
        "coordinates": [
          x
          x
        ],
        "type": "Point"
      }
    ],
    "google_workspace.drive.primary_event": [
      true
    ],
    "cloud.instance.id": [
      "1709224677170316971"
    ],
    "agent.type": [
      "filebeat"
    ],
    "event.module": [
      "google_workspace"
    ],
    "user.email": [
      "terrance@dejesusarcheology.com"
    ],
    "related.ip": [
      "2603:6010:ed00:b789:1860:feb:feee:41e8"
    ],
    "source.geo.country_iso_code": [
      "US"
    ],
    "source.user.id": [
      "115903088752625509360"
    ],
    "elastic_agent.snapshot": [
      false
    ],
    "user.domain": [
      "dejesusarcheology.com"
    ],
    "source.as.organization.name.text": [
      "TWC-10796-MIDWEST"
    ],
    "elastic_agent.id": [
      "f4011165-a1b8-4d8e-b902-56ba8835cc28"
    ],
    "google_workspace.drive.owner_is_team_drive": [
      false
    ],
    "data_stream.namespace": [
      "default"
    ],
    "file.name": [
      "Copy of 2023 Fiscal Year Report"
    ],
    "source.as.organization.name": [
      "TWC-10796-MIDWEST"
    ],
    "source.geo.continent_name": [
      "North America"
    ],
    "google_workspace.kind": [
      "admin#reports#activity"
    ],
    "event.action": [
      "copy"
    ],
    "event.ingested": [
      "2023-03-29T20:15:37Z"
    ],
    "@timestamp": [
      "2023-03-29T20:12:17.703Z"
    ],
    "cloud.account.id": [
      "elastic-security-dev"
    ],
    "google_workspace.drive.file.type": [
      "spreadsheet"
    ],
    "data_stream.dataset": [
      "google_workspace.drive"
    ],
    "agent.ephemeral_id": [
      "4c4fa27f-35e7-465b-8383-0b6e6cea1884"
    ],
    "source.user.domain": [
      "dejesusarcheology.com"
    ],
    "source.geo.country_name": [
      "United States"
    ],
    "event.id": [
      "-6699164353799691985"
    ],
    "event.dataset": [
      "google_workspace.drive"
    ],
    "user.name.text": [
      "terrance"
    ],
    "cloud.instance.name": [
      "ubuntu-server-tdejesus"
    ],
    "cloud.project.id": [
      "elastic-security-dev"
    ]
  }
}
```

OAuth permission granted.

```sql
{
  "_index": ".ds-logs-google_workspace.token-default-2023.03.28-000001",
  "_id": "YxVOYBW4evEaA0AE4fnLuxKstd0=",
  "_score": 1,
  "fields": {
    "elastic_agent.version": [
      "8.6.2"
    ],
    "source.user.email": [
      "terrance@dejesusarcheology.com"
    ],
    "cloud.availability_zone": [
      "us-east1-b"
    ],
    "google_workspace.token.scope.data.scope_name": [
      "https://mail.google.com/",
      "https://www.google.com/m8/feeds",
      "https://www.googleapis.com/auth/userinfo.email",
      "openid"
    ],
    "source.user.name.text": [
      "terrance"
    ],
    "source.ip": [
      "2603:6010:ed00:b789:1860:feb:feee:41e8"
    ],
    "agent.name": [
      "ubuntu-server-tdejesus"
    ],
    "event.agent_id_status": [
      "verified"
    ],
    "event.kind": [
      "event"
    ],
    "user.id": [
      "115903088752625509360"
    ],
    "cloud.instance.name.text": [
      "ubuntu-server-tdejesus"
    ],
    "input.type": [
      "httpjson"
    ],
    "data_stream.type": [
      "logs"
    ],
    "tags": [
      "forwarded",
      "google_workspace-token"
    ],
    "google_workspace.token.client.type": [
      "WEB"
    ],
    "related.user": [
      "115903088752625509360",
      "terrance",
      "terrance@dejesusarcheology.com"
    ],
    "event.provider": [
      "token"
    ],
    "cloud.provider": [
      "gcp"
    ],
    "cloud.machine.type": [
      "e2-medium"
    ],
    "agent.id": [
      "f4011165-a1b8-4d8e-b902-56ba8835cc28"
    ],
    "cloud.service.name": [
      "GCE"
    ],
    "ecs.version": [
      "8.6.0"
    ],
    "event.created": [
      "2023-03-29T22:00:14.289Z"
    ],
    "google_workspace.token.app_name": [
      "Untitled project"
    ],
    "organization.id": [
      "C00qtspd5"
    ],
    "agent.version": [
      "8.6.2"
    ],
    "related.hosts": [
      "dejesusarcheology.com"
    ],
    "source.user.name": [
      "terrance"
    ],
    "user.name": [
      "terrance"
    ],
    "cloud.instance.id": [
      "1709224677170316971"
    ],
    "agent.type": [
      "filebeat"
    ],
    "event.module": [
      "google_workspace"
    ],
    "related.ip": [
      "2603:6010:ed00:b789:1860:feb:feee:41e8"
    ],
    "user.email": [
      "terrance@dejesusarcheology.com"
    ],
    "source.user.id": [
      "115903088752625509360"
    ],
    "elastic_agent.snapshot": [
      false
    ],
    "user.domain": [
      "dejesusarcheology.com"
    ],
    "google_workspace.token.client.id": [
      "213527643582-pd407ig7fmfsr4if57mivv8u2q4kf7dm.apps.googleusercontent.com"
    ],
    "elastic_agent.id": [
      "f4011165-a1b8-4d8e-b902-56ba8835cc28"
    ],
    "google_workspace.token.scope.value": [
      "https://mail.google.com/",
      "https://www.google.com/m8/feeds",
      "https://www.googleapis.com/auth/userinfo.email",
      "openid"
    ],
    "data_stream.namespace": [
      "default"
    ],
    "google_workspace.kind": [
      "admin#reports#activity"
    ],
    "event.ingested": [
      "2023-03-29T22:00:15Z"
    ],
    "event.action": [
      "authorize"
    ],
    "@timestamp": [
      "2023-03-29T20:13:42.864Z"
    ],
    "google_workspace.etag": [
      "\"1rpbeKGlFXXrXfE6I4jSKiwRKeqWfVrVpA_lfqSodas/DCzW8-u_apAE_-TBOq-RfFrwleI\""
    ],
    "cloud.account.id": [
      "elastic-security-dev"
    ],
    "data_stream.dataset": [
      "google_workspace.token"
    ],
    "source.user.domain": [
      "dejesusarcheology.com"
    ],
    "agent.ephemeral_id": [
      "4c4fa27f-35e7-465b-8383-0b6e6cea1884"
    ],
    "event.id": [
      "6608422186174507988"
    ],
    "google_workspace.token.scope.data.product_bucket": [
      "GMAIL",
      "CONTACTS",
      "IDENTITY",
      "OTHER",
      "IDENTITY",
      "OTHER"
    ],
    "event.dataset": [
      "google_workspace.token"
    ],
    "user.name.text": [
      "terrance"
    ],
    "cloud.project.id": [
      "elastic-security-dev"
    ],
    "cloud.instance.name": [
      "ubuntu-server-tdejesus"
    ]
  }
}
```

</details>

### CODE REVIEW

File: rules/integrations/google_workspace/initial_access_google_workspace_object_copied_from_external_drive_and_access_granted_to_custom_application.toml [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+language = "eql"
+license = "Elastic License v2"
+name = "Google Workspace Object Copied from External Drive and Access Granted to Custom Application"
+note = """## Triage and analysis
+
+### Investigating Google Workspace Resource Copied from External Drive and Access Granted to Custom Application
+
+Google Workspace user's have the capability to share access to Drive objects such as documents, sheets, forms and more. Delivery of shared access can be done by email delivery or via shared link. Shared link URIs have `view` or `edit` in the parameters based on set permissions. The `copy` parameter is another option which allows the recipient to copy the object to their own Drive, thus the object has the privileges of the recipient. Specific objects allow container-bound scripts that leverage Google's Apps Script platform. Container-bound scripts can contain malicious code that executes with the recipient's privileges if in their Drive.
+
+This rule detections when a user copies an external Drive object to their Drive storage, followed by granting permissions to a custom application via OAuth prompt.
```
Comment (imays11) [inline rules/integrations/google_workspace/initial_access_google_workspace_object_copied_from_external_drive_and_access_granted_to_custom_application.toml:38]:
```suggestion
Google Workspace users can share access to Drive objects such as documents, sheets, and forms via email delivery or a shared link. Shared link URIs have parameters like `view` or `edit` to indicate the recipient's permissions. The `copy` parameter allows the recipient to copy the object to their own Drive, which grants the object with the same privileges as the recipient. Specific objects in Google Drive allow container-bound scripts that run on Google's Apps Script platform. Container-bound scripts can contain malicious code that executes with the recipient's privileges if in their Drive.

This rule aims to detect when a user copies an external Drive object to their Drive storage and then grants permissions to a custom application via OAuth prompt.
```
Just edited for clarity and readability, not all necessary changes but suggested.

File: rules/integrations/google_workspace/initial_access_google_workspace_object_copied_from_external_drive_and_access_granted_to_custom_application.toml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+#### Possible investigation steps
+- Identify user account(s) associated by reviewing `user.name` or `source.user.email` in the alert.
+- Identify the name of the file copied by reviewing `file.name` as well as the `file.id` for triaging.
+- Identify the file type by reviewing `google_workspace.drive.file.type`.
+- With the information gathered so far, query across data for the file metadata to determine if this activity is isolated or widespread.
+- Within the OAuth token event, identify the application name by reviewing `google_workspace.token.app_name`.
+    - Review the application ID as well from `google_workspace.token.client.id`.
+    - This metadata can be used to report the malicious application to Google for permanent blacklisting.
+- Review what permissions were granted access by the user by reviewing `google_workspace.token.scope.data.scope_name`.
```
Comment (imays11) [inline rules/integrations/google_workspace/initial_access_google_workspace_object_copied_from_external_drive_and_access_granted_to_custom_application.toml:48]:
```suggestion
- Identify the permissions granted to the application by the user by reviewing `google_workspace.token.scope.data.scope_name`.
```
suggested change for clarity and uniformity

File: rules/integrations/google_workspace/initial_access_google_workspace_object_copied_from_external_drive_and_access_granted_to_custom_application.toml (hunk 1 above)
Comment (terrancedejesus) [reply rules/integrations/google_workspace/initial_access_google_workspace_object_copied_from_external_drive_and_access_granted_to_custom_application.toml:38]: Thanks Isai!

**Review by imays11 — APPROVED** [review]
Looks good!

**Review by DefSecSentinel — APPROVED** [review]
LGTM

**Pull request merged**